File: prob1.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        if(nums == null || nums.length == 0){
            return new int[]{-1, -1};
        }
        int first = findfirst(nums, target);
        if(first == -1){
            return new int[]{-1, -1};
        }
        int last = findlast(nums, target);
        return new int[]{first, last};
    }
    private int findfirst(int[] nums,int target){
        int low = 0; 
        int high = nums.length-1;
        
        while(low <= high){
            int mid = low + (high -low)/2;
            if(nums[mid] == target){
                if(mid == 0 || nums[mid -1] != target){
                    return mid;
                }
                high = mid -1;
            } else if(target > nums[mid]){
                low = mid+1;
            }else{
                high = mid -1;
            }
        }
        return -1;
    }
    
    private int findlast(int[] nums, int target){
        int low = 0; 
        int high = nums.length-1;
        while(low <= high){
            int mid = low + (high - low)/2;
            if(nums[mid] == target){
                if(mid == nums.length-1 || nums[mid +1] != target){
                    return mid;
                }
                low = mid+1;
            }else if(target > nums[mid]){
                low = mid +1;
            }else{
                high = mid-1;
            }
    }
        return -1;
    }
}

File: prob1.py
# Time Complexity : O(log(n))
# Space Complexity :O(1)
# Passed on Leetcode: yes
# Approch: Apply Binary search, find the mid element, 
# if target is mid element the travers left and right from the mid to find the first and last occurance

class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        
        l = 0
        r = len(nums)-1
        op = [-1,-1]
        while l <= r:
            mid = (l+r)//2
            if target == nums[mid]:
                i = j = mid
                while i >= 0 and nums[i] == target:
                    op[0] = i
                    i -= 1
                while j < len(nums) and nums[j] == target:
                    op[1] = j
                    j += 1
            if target < nums[mid]:
                r = mid-1
            else:
                l = mid+1
                
        return op